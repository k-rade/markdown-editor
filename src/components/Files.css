.files {
    max-width: 20rem;
    width: 100%;
    height: 100vh;
    top: 0;
    transition: left 0.5s;
    transition: margin-left 0.35s;
    overflow-y: auto;
    border-left: 1px solid var(--color-gray-500);
    padding: 0 1.5rem 1.5rem;
}

.files-open {
    margin-left: 0;
}

.files-header {
    height: 5rem;
    display: flex;
    align-items: center;
    justify-content: space-between;
    font-weight: 500;
    font-size: 1.25rem;
    border-bottom: 1px solid var(--color-gray-500);
}

.files-button {
    position: relative;
}

.files-list {
    display: flex;
    flex-direction: column;
    gap: 1rem;
    margin-top: 1rem;
}

.file {
    color: var(--color-gray-900);
    z-index: 1;
}

.file-header {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.file-active .file-name {
    color: var(--color-white-one);
    text-decoration: underline;
}

.file-active .file-content {
    color: var(--color-white-three);
}

.file-name {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    max-width: 14rem;
    cursor: pointer;
    color: var(--color-white-three);
    font-weight: 600;
    flex: 1;
}

.file-name:hover {
    text-decoration: underline;
}

.file-content {
    height: 2.5rem;
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    font-size: 0.75rem;
    cursor: pointer;
}

.file-text {
    color: var(--color-gray-900);
    margin: 1.25rem;
}


@media screen and (min-width:1024px) {
    .files-close {
        display: none;
    }
}

@media screen and (max-width:1024px) {
    .files {
        right: -100%;
        position: fixed;
        background-color: var(--color-gray-100);
        transition: right 0.5s;
        height: calc(100% - 4rem);
    }

    .files-open {
        right: 0;
    }

    .files-overlay {
        z-index: 0;
        top: 0;
        left: 0;
        position: fixed;
        display: none;
        width: 100vw;
        height: 100vh;
        background-color: hsla(214, 8%, 8%, 0.85);
    }

    .files-overlay-visible {
        display: block;
    }
}