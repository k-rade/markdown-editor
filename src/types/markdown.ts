export interface MarkdownFile {
  name: string;
  markdown: string;
}
