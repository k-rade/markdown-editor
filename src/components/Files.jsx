import { useState } from "react";
import IconButton from "./IconButton";
import { ReactComponent as CloseIcon } from "../assets/icons/x.svg";
import { ReactComponent as DeleteIcon } from "../assets/icons/trash.svg";
import { ReactComponent as FolderIcon } from "../assets/icons/folder.svg";

import "./Files.css";

const Files = ({ files, selectFile, deleteFile }) => {
  const [showFiles, setShowFiles] = useState(false);

  return (
    <>
      <div className="files-button">
        {files.length > 0 && <div className="files-amount">{files.length}</div>}
        <IconButton
          icon={<FolderIcon />}
          onClick={() => setShowFiles((prevState) => !prevState)}
        />
      </div>
    </>
  );
};

export default Files;
