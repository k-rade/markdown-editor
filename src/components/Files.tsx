import { MarkdownFile } from "../types/markdown";
import { FiFilePlus, FiX } from "react-icons/fi";
import File from "./File";

interface Props {
  files: MarkdownFile[];
  currentFile?: MarkdownFile;
  selectFile: (fileName: string) => void;
  deleteFile: (fileName: string) => void;
  toggleFilesShow: () => void;
  showFiles: boolean;
}

const Files = ({
  files,
  currentFile,
  selectFile,
  deleteFile,
  toggleFilesShow,
  showFiles,
}: Props): JSX.Element => {
  const handleFilesSelect = (name: string): void => {
    selectFile(name);
    toggleFilesShow();
  };

  return (
    <>
      {showFiles && (
        <div
          data-testid="files-overlay"
          onClick={toggleFilesShow}
          className="z-10 ml-20 lg:ml-0 lg:hidden fixed inset-0 w-full h-full bg-zinc-900/70"
        ></div>
      )}
      <div
        className={`${
          showFiles ? "left-0 sm:ml-20" : "-left-full "
        } overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-800 dark:border-zinc-700 w-full max-w-sm`}
      >
        <div className="border-b h-16 flex items-center justify-between px-6 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-700">
          <div>
            Notes{" "}
            <span className="text-sm text-zinc-400 font-normal ml-2">
              {`${files.length} ${files.length === 1 ? "File" : "Files"}`}
            </span>
          </div>
          <FiX
            data-testid="x-icon"
            onClick={toggleFilesShow}
            size="1.375rem"
            className="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
          />
        </div>
        {files.length ? (
          <div className="w-full max-w-sm">
            {files.map(
              (file, i): JSX.Element => (
                <File
                  key={i}
                  currentFileName={currentFile?.name}
                  file={file}
                  deleteFile={deleteFile}
                  selectFile={handleFilesSelect}
                />
              )
            )}
          </div>
        ) : (
          <div className="flex flex-col items-center justify-center gap-1 mt-8">
            <FiFilePlus
              size="2.5rem"
              className="stroke-1 stroke-zinc-300 mb-1"
            />
            <div className="text-zinc-700 font-semibold">No notes saved</div>
            <div className="text-sm text-zinc-400">
              Get started by saving your first note.
            </div>
          </div>
        )}
      </div>
    </>
  );
};

export default Files;
