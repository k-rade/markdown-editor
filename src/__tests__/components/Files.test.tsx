import { fireEvent, render, RenderResult } from "@testing-library/react";
import Files from "../../components/Files";
import { MarkdownFile } from "../../types/markdown";

describe("Files", (): void => {
  const mockFiles = [
    { name: "File 1", markdown: "content" },
    { name: "File 2", markdown: "" },
  ] as MarkdownFile[];

  const mockSelectFile = jest.fn();
  const mockDeleteFile = jest.fn();
  const mockToggleFilesShow = jest.fn();

  const renderTestComponent = (
    files = mockFiles,
    showFiles = false
  ): RenderResult =>
    render(
      <Files
        files={files}
        currentFile={mockFiles[0]}
        selectFile={mockSelectFile}
        deleteFile={mockDeleteFile}
        toggleFilesShow={mockToggleFilesShow}
        showFiles={showFiles}
      />
    );

  it("should render", (): void => {
    const { baseElement } = renderTestComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it("should render with overlay", (): void => {
    const { baseElement } = renderTestComponent(mockFiles, true);

    expect(baseElement).toMatchSnapshot();
  });

  it("should render without files", (): void => {
    const { baseElement } = renderTestComponent([]);

    expect(baseElement).toMatchSnapshot();
  });

  it("should call selectFile on file name", (): void => {
    const { getByText } = renderTestComponent();

    fireEvent.click(getByText("File 1"));

    expect(mockSelectFile).toHaveBeenNthCalledWith(1, "File 1");
  });

  it("should call selectFile on file content", (): void => {
    const { getByText } = renderTestComponent();

    fireEvent.click(getByText("content"));

    expect(mockSelectFile).toHaveBeenNthCalledWith(1, "File 1");
  });

  it("should call deleteFile", (): void => {
    const { getAllByTestId } = renderTestComponent();

    fireEvent.click(getAllByTestId("trash-icon")[0]);

    expect(mockDeleteFile).toHaveBeenCalledTimes(1);
  });

  it("should call toggleFilesShow", (): void => {
    const { getByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("x-icon"));

    expect(mockToggleFilesShow).toHaveBeenCalledTimes(1);
  });
});
