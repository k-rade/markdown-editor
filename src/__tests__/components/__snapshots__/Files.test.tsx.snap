// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files should render 1`] = `
<body>
  <div>
    <div
      class="-left-full  overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-800 dark:border-zinc-700 w-full max-w-sm"
    >
      <div
        class="border-b h-16 flex items-center justify-between px-4 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-700"
      >
        <div>
          Notes
           
          <span
            class="text-sm text-zinc-400 font-normal ml-1"
          >
            2 Files
          </span>
        </div>
        <svg
          class="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
          data-testid="x-icon"
          fill="none"
          height="1.375rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.375rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="w-full max-w-sm"
      >
        <div
          class="bg-zinc-50 dark:bg-zinc-700/25 border-b p-4 dark:border-zinc-700"
        >
          <p
            class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
          >
            <span>
              File 1
            </span>
            <svg
              class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
              data-testid="trash-icon"
              fill="none"
              height="1.125rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.125rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </p>
          <p
            class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
          >
            content
          </p>
        </div>
        <div
          class=" border-b p-4 dark:border-zinc-700"
        >
          <p
            class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
          >
            <span>
              File 2
            </span>
            <svg
              class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
              data-testid="trash-icon"
              fill="none"
              height="1.125rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.125rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </p>
          <p
            class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
          >
            Empty note
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Files should render with overlay 1`] = `
<body>
  <div>
    <div
      class="z-10 ml-20 lg:ml-0 lg:hidden fixed inset-0 w-full h-full bg-zinc-900/70"
      data-testid="files-overlay"
    />
    <div
      class="left-0 sm:ml-20 overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-800 dark:border-zinc-700 w-full max-w-sm"
    >
      <div
        class="border-b h-16 flex items-center justify-between px-4 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-700"
      >
        <div>
          Notes
           
          <span
            class="text-sm text-zinc-400 font-normal ml-1"
          >
            2 Files
          </span>
        </div>
        <svg
          class="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
          data-testid="x-icon"
          fill="none"
          height="1.375rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.375rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="w-full max-w-sm"
      >
        <div
          class="bg-zinc-50 dark:bg-zinc-700/25 border-b p-4 dark:border-zinc-700"
        >
          <p
            class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
          >
            <span>
              File 1
            </span>
            <svg
              class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
              data-testid="trash-icon"
              fill="none"
              height="1.125rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.125rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </p>
          <p
            class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
          >
            content
          </p>
        </div>
        <div
          class=" border-b p-4 dark:border-zinc-700"
        >
          <p
            class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
          >
            <span>
              File 2
            </span>
            <svg
              class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
              data-testid="trash-icon"
              fill="none"
              height="1.125rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.125rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </p>
          <p
            class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
          >
            Empty note
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Files should render without files 1`] = `
<body>
  <div>
    <div
      class="-left-full  overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-800 dark:border-zinc-700 w-full max-w-sm"
    >
      <div
        class="border-b h-16 flex items-center justify-between px-4 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-700"
      >
        <div>
          Notes
           
          <span
            class="text-sm text-zinc-400 font-normal ml-1"
          >
            0 Files
          </span>
        </div>
        <svg
          class="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
          data-testid="x-icon"
          fill="none"
          height="1.375rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.375rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="flex flex-col items-center justify-center gap-1 mt-8"
      >
        <svg
          class="stroke-1 stroke-zinc-300 mb-1"
          fill="none"
          height="2.5rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="2.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
          />
          <polyline
            points="14 2 14 8 20 8"
          />
          <line
            x1="12"
            x2="12"
            y1="18"
            y2="12"
          />
          <line
            x1="9"
            x2="15"
            y1="15"
            y2="15"
          />
        </svg>
        <div
          class="text-zinc-700 font-semibold"
        >
          No notes saved
        </div>
        <div
          class="text-sm text-zinc-400"
        >
          Get started by saving your first note.
        </div>
      </div>
    </div>
  </div>
</body>
`;
